examples/fluids/src/dm_utils.c

```diff
@@ -96,6 +47,202 @@ PetscErrorCode DMFieldToDSField(DM dm, DMLabel domain_label, PetscInt dm_field,
+
+  PetscFunctionBeginUser;
+  PetscCall(
+      DMPlexGetLocalOffsets(dm, domain_label, label_value, height, dm_field, &num_elem, &elem_size, &num_comp, &num_dof, &restriction_offsets_petsc));
+  PetscCall(IntArrayP2C(num_elem * elem_size, &restriction_offsets_petsc, &restriction_offsets_ceed));
+  PetscCallCeed(ceed, CeedElemRestrictionCreate(ceed, num_elem, elem_size, num_comp, 1, num_dof, CEED_MEM_HOST, CEED_COPY_VALUES,
+                                                restriction_offsets_ceed, restriction));
+  PetscCall(PetscFree(restriction_offsets_ceed));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/**
+  @brief Create `CeedElemRestriction` from `DMPlex` domain for mesh coordinates.
+
+  Not collective across MPI processes.
+
+  @param[in]   ceed          `Ceed` context
+  @param[in]   dm            `DMPlex` holding mesh
+  @param[in]   domain_label  Label for `DMPlex` domain
+  @param[in]   label_value   Stratum value
+  @param[in]   height        Height of `DMPlex` topology
+  @param[out]  restriction   `CeedElemRestriction` for mesh
+
+  @return An error code: 0 - success, otherwise - failure
+**/
+PetscErrorCode DMPlexCeedElemRestrictionCoordinateCreate(Ceed ceed, DM dm, DMLabel domain_label, PetscInt label_value, PetscInt height,
+                                                         CeedElemRestriction *restriction) {
+  DM        dm_coord;
+  PetscBool is_simplex = PETSC_FALSE;
+
+  PetscFunctionBeginUser;
+  PetscCall(DMGetCellCoordinateDM(dm, &dm_coord));
+  if (!dm_coord) {
+    PetscCall(DMGetCoordinateDM(dm, &dm_coord));
+  }
+  PetscCall(DMPlexIsSimplex(dm_coord, &is_simplex));
+  if (!is_simplex) PetscCall(DMPlexSetClosurePermutationTensor(dm_coord, PETSC_DETERMINE, NULL));
+  PetscCall(DMPlexCeedElemRestrictionCreate(ceed, dm_coord, domain_label, label_value, height, 0, restriction));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/**
+  @brief Create `CeedElemRestriction` from `DMPlex` domain for auxilury `QFunction` data.
+
+  Not collective across MPI processes.
+
+  @param[in]   ceed           `Ceed` context
+  @param[in]   dm             `DMPlex` holding mesh
+  @param[in]   domain_label   Label for `DMPlex` domain
+  @param[in]   label_value    Stratum value
+  @param[in]   height         Height of `DMPlex` topology
+  @param[in]   q_data_size    Number of components for `QFunction` data
+  @param[in]   is_collocated  Boolean flag indicating if the data is collocated on the nodes (`PETSC_TRUE`) on on quadrature points (`PETSC_FALSE`)
+  @param[out]  restriction    Strided `CeedElemRestriction` for `QFunction` data
+
+  @return An error code: 0 - success, otherwise - failure
+**/
+static PetscErrorCode DMPlexCeedElemRestrictionStridedCreate(Ceed ceed, DM dm, DMLabel domain_label, PetscInt label_value, PetscInt height,
+                                                             PetscInt q_data_size, PetscBool is_collocated, CeedElemRestriction *restriction) {
+  PetscInt num_elem, num_qpts, dm_field = 0;
```

Comment:
I'm not sure how I feel about always setting `dm_field` equal to 0. On one hand, I don't think I have an example of needing strided restrictions for DM fields other than the default one, but I _do_ create restrictions for a single field of a DM. 

I don't see any reason to not ask for `dm_field` here. @jeremylt was there any specific thought given to this, or was it just that Ratel doesn't use anything other than `dm_field = 0` and so reducing the surface area of the API?

Reply:
The quadrature space has to have the same number of points across all fields, even if each field has different numbers of nodes on each element

Reply:
> The quadrature space has to have the same number of points across all fields, even if each field has different numbers of nodes on each element

That's true only if they're used in the same qfunctions though, right? I can't think of a reason to have two fields in the same DM, used in different QFunctions with different quadrature rules, but I've had a field be used in one QFunction, then used in a wholly unrelated QFunction before (in the SmartSim PR, I use a field to perform differential filtering, then use that field to gather the ML training data together). 

All that said, there might be reason to, but that's very complicated and if I run into it, running the plumbing for adding a `dm_field` shouldn't be too hard.

Reply:
The convention I've been using is 1 dm == 1 quadrature space, especially since the coordinates can only have one quadrature space. I agree we should see if we have a need to break this convention in the future and generalize further then.